### DISCUSSION TITLE
Linting/Formatting/Type Checking with PEP 582 in VS Code

### DISCUSSION THREAD

**Discussion by acalvino4** [pdm-project/pdm#1138] [Q&A] [answered]
I know VSCode's python extension has yet to achieve substantial support for PEP 582, but I am wondering if anyone has any tips or ideas about getting the in-editor error/warning highlights working when using various code quality tools.

So far I have installed `black` and formatting the files works fine when I run it as a command, but in my editor there is no indication when I have a formatting problem. How might I enable this?

<img width="298" alt="image" src="https://user-images.githubusercontent.com/43254861/173863509-47cd2462-7706-4b34-8dd6-4dcf0e064053.png">
 ^ inconsistent quote usage, yet no warning

Here is my `.vscode/settings.json`

```
{
  "python.autoComplete.extraPaths": ["__pypackages__/3.10/lib"],
  "python.analysis.extraPaths": ["__pypackages__/3.10/lib"],
  "python.formatting.provider": "black"
}
```

**Comment by frostming** [comment, marked as answer]
Do you have autoformat config?

black and linters like flake8 are all *static* tools, they don't rely on any dependency of the project. So PEP 582 won't be a problem.

> but in my editor there is no indication

What kind of indication? black doesn't show indicators on format problems, AFAIK.

**Reply by pawamoy** [reply]
I myself use flake8-black, and configure VSCode to find my flake8 config file, so that I get Black's warning in the IDE :smile:

**Reply by acalvino4** [reply]
@frostming I'm coming from a js background, in which prettier is the equivalent of black. I get errors highlighted there so I assume (assuming the vscode integration is setup correctly) that it would be the same for black. Come to think of it however, I wonder if my prettier errors only show bc I have it set up as an eslint plugin. If that is the case, it makes perfect sense that @pawamoy has it working with his linter. I will give this a try and report back.

**Reply by acalvino4** [reply]
well, using it as a flake8 plugin worked.

I have to say though, after living with eslint, I am a bit disappointed by the vagueness of the messages at times ("Black would make changes"), the lack of links to docs for specific errors in the hover menu, the fact that I need to explicitely save before the errors are updated, and the lack of auto-fix options. If you know of solutions to any of those problem, let me know, but I'll mark this as answered.

**Reply by pawamoy** [reply]
> vagueness of the messages

True, the message is not helpful at all. Though you're simply supposed to run Black, you don't really need to know why Black isn't happy (but that would be better of course). I guess Black just doesn't give this kind of details, so the flake8-black plugin cannot do better 😕 

> lack of links to docs for specific errors

Yeah that's annoying, it would help so much to get links to the explanation of the rules. It probably requires a coordinated effort from the IDE and the linter/linter plugins to achieve this.

> the fact that I need to explicitely save before the errors are updated

I think that's just how VSCode works? Running external programs on each keystroke, even with a minimal delay, would cost too much I gues? Maybe there's a VSCode setting to alter that?

> lack of auto-fix options

Auto-fix would be great as well 😍 Or at least auto-add-noqa-comments. Some cases could "easily" be auto-fixed, but I guess that would require some kind of support from flake8 itself. Or that could be implemented in autoflake or something.

**Reply by acalvino4** [reply]
Eslint shows errors almost instantaneously on the open file, even before saving, but I'm guessing the start up time of the python interpreter makes that infeasible for flake8, as you suggest. We really are spoiled by the quality and unity of tooling in the js world, and I look forward to python hopefully catching up some day.
